### ISSUE TITLE
Duplicate files while scanning library

### ISSUE THREAD

**Issue by peczony** [cmus/cmus#423]
While scanning my library **in a symlinked folder** cmus constantly finds duplicate files, like this:

![](https://i.imgur.com/Fpdi61Z.png)

Sometimes there are only two of them, other times three or even four (like in this example).

I probably do have some duplicate files, but it finds duplicates for practically every file, which is impossible — I have searched for many of supposedly duplicate files and found only one of each. I am unsure what info I should provide to clarify the issue :(, so please ask questions if you care to help.

```
Darwin Kernel Version 15.3.0, cmus v2.7.1
```

On the other note, how am I supposed to deal with duplicates anyway? I read the man, but couldn't find any way to display, e.g. the file path of the selected file (in pane 1). Duplicates make listening in album mode difficult. Have you considered providing the user with some option to deal with them, e.g. delete the duplicates, or add an additional level of grouping (e.g file path) that would remove the duplicates?

**Comment by z3t0** [comment]
You can delete individual songs with the "D" key

**Comment by z3t0** [comment]
It would be a good idea to add a feature where duplicate files are not added

**Comment by peczony** [comment]
@z3t0 Wouldn’t they be re-added if I add new music to library and then refresh cache?

Anyway, looks like I found the original problem — symlinked folder was the culprit. If I added the folder by its real path, obsolete duplicates didn’t occur.

Still unsure how to deal with legitimate duplicates though.

**Comment by flyingmutant** [comment]
What do you mean by «legitimate duplicates»? Files with identical content, but different paths?

**Comment by peczony** [comment]
E.g. a band has a single and an album with the same name. They will be grouped together, which is not very convenient.

I think that third level of grouping (by folder path) would solve this problem.

**Comment by jolange** [comment]
But that would impose a folder structure (Artist/Album/Track) which I think you should definitely not do.
If there are different albums/singles named the same for one artist, then I think that's the actual problem. I personally would try to make this clear from the metadata and call them "Name" and "Name (Single)".

**Comment by peczony** [comment]
I do not think this approach *imposes* anything: the user is free to choose whatever folder structure suits him better. I propose that the path/folder structure be used only as a last means for disambiguation, in the cases where all metadata for files is identical. It is also convenient because file paths are already stored in the library, so you do not need to cache any more data.

Before grouping by folder path, other levels of grouping may be added to solve the same problem, e.g musicbrainz ID. However, this approach has the aforementioned disadvantage of needing to cache more data.

**Comment by jolange** [comment]
Ok, but in my opinion "conditional disambiguation" is not desirable.
Apart from that, using the location for distinction purposes goes a bit against the idea of a media library.

**Comment by peczony** [comment]
> in my opinion "conditional disambiguation" is not desirable

Could you please elaborate why? I am genuinely interested. I understand the controversial ideological grounds, but I can’t see any drawbacks from an UX perspective.

**Comment by jolange** [comment]
Well, thats a purley personal view, but I like to have one defined behaviour (yes, you could argue that this is still well-defined).
I would certainly be at least surprised (if not confused) if suddenly there was a further grouping in one place that I don't see anywhere else.
And again: I understand your point, but I think the disambiguation should explicitly happen in the album title (an attribute belonging to a media library item) rather than using file paths (which a library is trying to abstract from).

Just my two cents.

**Comment by woffs** [comment]
I stumbled over the same problem. I have a directory tree with some duplicated (symlinked) content. For instance I have a subdir to be copied to a mobile MP3 player and another subdir with ripped CDs, and some of the content is in both subdirs, and it looks like the picture above. Would be nice if there was an option to squeeze the duplicated entries simply together somehow (if all metadata and filesize is identical and only the path is different). Otherwise I could try to reorganize my tree and make a cmus-subtree with no duplicated content. ;)

**Comment by mahkoh** [comment]
How do other players handle this?

**Comment by nefthy** [comment]
This can be fixed by calling <code>realpath</code> on the filename before processing. <code>realpath</code> will resolve symlinks and relative paths to a canonical absolute paths.

**Comment by forgottenswitch** [comment]
https://github.com/cmus/cmus/pull/556 would group symlinked directories into discrete albums.
But if there is one or more duplicate in the same dir (file+symlink or symlink+symlink),
then only the `realpath` or metadata-based (or file hash-based) deduplication would help.

**Comment by forgottenswitch** [comment]
If you have symlinks to *directories*, https://github.com/cmus/cmus/pull/556 would resolve this by putting files in each into discrete albums.
If there are symlinks to *files* (in the same dir), then deduplication is necessary.
(be it `realpath`, metadata, or file hash).

**Comment by forgottenswitch** [comment]
https://github.com/cmus/cmus/pull/556 would treat a dir and a symlink as two separate albums
(acting the "duplicate if user so desires" way :)

**Comment by xeverse** [comment]
What about this realpath anyway

**Comment by simbalion** [comment]
I just read this whole thread and with all due respect it's nonsense. If you've got a large library of music manually deleting duplicates is a pain and they'll be re-added on the next library re-fresh. 

I'm not going to presume to tell any of you what the right solution is, but the right result is that the Artist / Album view should not be displaying duplicates under any circumstances other than if there are actually duplicate files in the directory. Anything else is broken.

**Comment by Erhannis** [comment]
I've used cmus for maybe a year, and I think this issue is my primary complaint.  I have albums I can't really listen to because each song plays twice, sometimes three times.  Perhaps an option that, when true, condenses by SHA256 hash all songs in each album?  (I.e., only one file is shown out of those that share its SHA256 hash.)  Granted, there might be cases where there's a duplicate song encoded differently, which this solution does not address.  Perhaps the option could have multiple possible values?  ("group_by_hash", "group_by_title", etc.)

**Comment by gardelin** [comment]
I also had this issue but in my case duplicated songs appear when I renamed music directory.
Fastest solution for me was hold down `D` key (delete command) in artist tab, and then re-add music folder
`:add ~/Music`

**Comment by kuzyn** [comment]
I have been using cmus for a few years now, and this is still a "thing". Any movement on this? @Erhannis  a hashtable is not a bad idea imo

**Comment by rachmadaniHaryono** [comment]
i have this issue and because there is no solution yet except #1052, i modify the cmus lib directly (`~/.config/cmus.lib.pl`)

the drawback is i lose the playcount. 

currently there is no way to export play count as of [^1] G

[^1]: https://github.com/cmus/cmus/issues/489

**Comment by tehmasterer** [comment]
Same issue here. I have a symlink in `~/Music` that points to my network share that holds all my music. When adding the symlink, cmus created duplicates of all the files. I have a lot of music, and sorting through them to delete is not feasible.

**Comment by rachmadaniHaryono** [comment]
i have to write my own python script to replace symlink

for example `/mnt/uuid/music/1.mp3` is equal to `/home/username/music/1.mp3`

```python
from pathlib import Path
import difflib
print_diff = True
start_text = '/mnt/uuid'

lib_path = Path('~/.config/cmus/lib.pl').expanduser()
text = lib_path.read_text()
lines = text.splitlines()
good, bad = [], []
for x in lines:
    (bad, good)[not x.startswith(start_text)].append(x)
new_lines = list(sorted(set(good +  [('/home/username' / Path(x).relative_to(start_text)).as_posix() for x in bad])))
if print_diff:
    print('\n'.join(list(difflib.context_diff(lines, new_lines))))
lib_path.write_text('\n'.join(new_lines))
```

**Comment by gavtroy** [comment]
I've noticed a flaw in the `ignore_duplicates` option I added; that cmus won't catch duplicates if the original is currently hidden by a filter. They'll at least be stripped out of the library again if cmus is restarted with the filter disabled. Unfortunately I can't think of a good way to fix it right now, aside from something extremely invasive like https://github.com/gavtroy/cmus/commit/filter-keeps-objects.
